internal/adapters/dataproviders/database/mongo/common.go

```diff
@@ -74,3 +78,200 @@ func (c *Connection) CheckConnection(ctx context.Context) bool {
+	Ctx                context.Context
+	Conn               *Connection
+	StartDate          time.Time
+	EndDate            time.Time
+	QuoteCollection    string
+	RetainedCollection string
+}
+
+func ListQuotesByDateRange[Q any, R QuoteHashProvider](
+	query QuoteQuery,
+	mapper func(bson.D) Q,
+) QuoteResult[Q, R] {
+	dbCtx, cancel := context.WithTimeout(query.Ctx, query.Conn.timeout)
+	defer cancel()
+	quotes, quoteHashes, err := fetchQuotesByDateRange(dbCtx, query.Conn, query.StartDate, query.EndDate, query.QuoteCollection, mapper)
+	if err != nil {
+		return QuoteResult[Q, R]{Error: err}
+	}
+	quoteHashToIndex := make(map[string]int, len(quoteHashes))
+	for i, hash := range quoteHashes {
+		if hash != "" {
+			quoteHashToIndex[hash] = i
+		}
+	}
+	retainedQuotes, additionalHashes, err := fetchRetainedQuotes[R](dbCtx, query.Conn, query.StartDate, query.EndDate, query.RetainedCollection, quoteHashes)
+	if err != nil {
+		return QuoteResult[Q, R]{Error: err}
+	}
+	if len(additionalHashes) > 0 {
+		additionalQuotes, additionalHashIndices, err := fetchAdditionalQuotes(dbCtx, query.Conn, query.QuoteCollection, additionalHashes, mapper)
+		if err != nil {
+			log.Errorf("Error processing additional quotes: %v", err)
+		} else {
+			baseIndex := len(quotes)
+			for i, hash := range additionalHashIndices {
+				if hash != "" {
+					quoteHashToIndex[hash] = baseIndex + i
+				}
+			}
+			quotes = append(quotes, additionalQuotes...)
+		}
+	}
+	logDbInteraction(Read, fmt.Sprintf("Found %d quotes and %d retained quotes in date range",
+		len(quotes), len(retainedQuotes)))
+	return QuoteResult[Q, R]{
+		Quotes:           quotes,
+		RetainedQuotes:   retainedQuotes,
+		QuoteHashToIndex: quoteHashToIndex,
+		Error:            nil,
+	}
+}
+
+func fetchQuotesByDateRange[Q any](
+	ctx context.Context,
+	conn *Connection,
+	startDate, endDate time.Time,
+	collectionName string,
+	mapper func(bson.D) Q,
+) ([]Q, []string, error) {
+	quoteFilter := bson.D{
+		{Key: "agreement_timestamp", Value: bson.D{
+			{Key: "$gte", Value: startDate.Unix()},
+			{Key: "$lte", Value: endDate.Unix()},
+		}},
+	}
+	var storedQuotes []bson.D
+	quoteCursor, err := conn.Collection(collectionName).Find(ctx, quoteFilter)
+	if err != nil {
+		return nil, nil, err
+	}
+	if err = quoteCursor.All(ctx, &storedQuotes); err != nil {
+		return nil, nil, err
+	}
+	quoteHashes := make([]string, 0, len(storedQuotes))
+	quotes := make([]Q, 0, len(storedQuotes))
+	for _, stored := range storedQuotes {
+		quoteObj := mapper(stored)
+		quotes = append(quotes, quoteObj)
+		hashValue, ok := getStringValueFromBSON(stored, "hash")
+		if ok {
+			quoteHashes = append(quoteHashes, hashValue)
+		}
+	}
+	return quotes, quoteHashes, nil
+}
+
+func getStringValueFromBSON(doc bson.D, key string) (string, bool) {
+	data, err := bson.Marshal(doc)
+	if err != nil {
+		return "", false
+	}
+	rawValue := bson.Raw(data).Lookup(key)
+	return rawValue.StringValueOK()
+}
+
+type QuoteHashProvider interface {
```

Comment:
the same interface that you created in the use case, this is why it should be defined at entity layer